src/main/java/htsjdk/samtools/CachingBAMFileIndex.java

```diff
@@ -143,31 +144,35 @@ public BAMFileSpan getSpanOverlapping(final Bin bin) {
      */
     @Override
     protected BAMIndexContent getQueryResults(final int referenceIndex) {
-        // WeakHashMap is a bit weird in that its lookups are done via equals() equality, but expirations must be
-        // handled via == equality.  This implementation jumps through a few hoops to make sure that == equality still
-        // holds even in the context of boxing/unboxing.
 
         // If this query is for the same reference index as the last query, return it.
-        if(mLastReferenceRetrieved!=null && mLastReferenceRetrieved == referenceIndex)
-            return mQueriesByReference.get(referenceIndex);
-
-        // If not, check to see whether it's available in the cache.
-        BAMIndexContent queryResults = mQueriesByReference.get(referenceIndex);
-        if(queryResults != null) {
-            mLastReferenceRetrieved = referenceIndex;
-            mQueriesByReference.put(referenceIndex,queryResults);
-            return queryResults;
+        if(lastReferenceIndex!=null && lastReferenceIndex == referenceIndex){
+            if(lastReference != null) {
+                cacheHits++;
+                return lastReference;
+            } else {
+                throw new RuntimeException("Expected to find a cached reference but we found null.  This is a bug.");
+            }
         }
 
-        // If not in the cache, attempt to load it from disk.
-        queryResults = query(referenceIndex,1,-1);
+        // If not attempt to load it from disk.
+        BAMIndexContent queryResults = query(referenceIndex,1,-1);
```

Comment:
When `query()` returns null, do you want to clear out the cache?

Reply:
If you do make this change, be sure to add a test, and document what it means to cache an "empty" result.